pkg/scheduler/tests/scheduler_operation_test.go

```diff
@@ -0,0 +1,514 @@
+// then we submit an app, the app tasks will be pending; then we add a
+// node to the cluster, then we see the app gets the allocation it needed.
+func TestSchedulerWithoutNodes(t *testing.T) {
+	// Register RM
+	configData := `
+partitions:
+  -
+    name: default
+    queues:
+      - name: root
+        submitacl: "*"
+        queues:
+          - name: a
+            resources:
+              guaranteed:
+                memory: 100
+                vcore: 10
+              max:
+                memory: 150
+                vcore: 20
+`
+	// Start all tests
+	ms := &mockScheduler{}
+	defer ms.Stop()
+
+	err := ms.Init(configData, false)
+	if err != nil {
+		t.Fatalf("RegisterResourceManager failed: %v", err)
+	}
+
+	// Check queues of cache and scheduler.
+	partitionInfo := ms.clusterInfo.GetPartition("[rm:123]default")
+	assert.Assert(t, nil == partitionInfo.Root.GetMaxResource(), "partition info max resource nil")
+
+	// Check scheduling queue root
+	schedulerQueueRoot := ms.getSchedulingQueue("root")
+	assert.Assert(t, nil == schedulerQueueRoot.QueueInfo.GetMaxResource())
+
+	// Check scheduling queue a
+	schedulerQueueA := ms.getSchedulingQueue("root.a")
+	assert.Assert(t, 150 == schedulerQueueA.QueueInfo.GetMaxResource().Resources[resources.MEMORY])
+
+	// Add one application
+	err = ms.proxy.Update(&si.UpdateRequest{
+		NewApplications: newAddAppRequest(map[string]string{"app-1": "root.a"}),
+		RmID:            "rm:123",
+	})
+
+	// Application should be accepted
+	ms.mockRM.waitForAcceptedApplication(t, "app-1", 1000)
+
+	// Check scheduling app
+	schedulingApp := ms.getSchedulingApplication("app-1")
+	assert.Equal(t, schedulingApp.ApplicationInfo.ApplicationID, "app-1")
+	assert.Equal(t, len(schedulingApp.ApplicationInfo.GetAllAllocations()), 0)
+
+	// App asks for 2 allocations
+	err = ms.proxy.Update(&si.UpdateRequest{
+		Asks: []*si.AllocationAsk{
+			{
+				AllocationKey: "alloc-1",
+				ResourceAsk: &si.Resource{
+					Resources: map[string]*si.Quantity{
+						"memory": {Value: 10},
+						"vcore":  {Value: 1},
+					},
+				},
+				MaxAllocations: 2,
+				ApplicationID:  "app-1",
+			},
+		},
+		RmID: "rm:123",
+	})
+
+	if err != nil {
+		t.Fatalf("UpdateRequest 2 failed: %v", err)
+	}
+
+	waitForPendingQueueResource(t, schedulerQueueA, 20, 1000)
+	waitForPendingQueueResource(t, schedulerQueueRoot, 20, 1000)
+	waitForPendingAppResource(t, schedulingApp, 20, 1000)
+
+	// no nodes available, no allocation can be made
+	ms.scheduler.MultiStepSchedule(16)
+
+	// pending resources should not change
+	waitForPendingQueueResource(t, schedulerQueueA, 20, 1000)
+	waitForPendingQueueResource(t, schedulerQueueRoot, 20, 1000)
+	waitForPendingAppResource(t, schedulingApp, 20, 1000)
+
+	// Register a node
+	err = ms.proxy.Update(&si.UpdateRequest{
+		NewSchedulableNodes: []*si.NewNodeInfo{
+			{
+				NodeID: "node-1:1234",
+				Attributes: map[string]string{
+					"si.io/hostname": "node-1",
+					"si.io/rackname": "rack-1",
```

Comment:
unused in tests should be removed and made an empty map

Reply:
done